pkl-core/src/main/java/org/pkl/core/module/ModuleKeyFactories.java

```diff
@@ -225,4 +250,45 @@ private static class FromServiceProviders {
       INSTANCE = Collections.unmodifiableList(factories);
     }
   }
+
+  /** Represents a module from an external reader process. */
+  private static final class External implements ModuleKeyFactory {
+    private final String scheme;
+    private final ExternalReaderProcess process;
+    private final long evaluatorId;
+    private ExternalModuleResolver resolver;
+
+    public External(String scheme, ExternalReaderProcess process, long evaluatorId) {
+      this.scheme = scheme;
+      this.process = process;
+      this.evaluatorId = evaluatorId;
+    }
+
+    private ExternalModuleResolver getResolver() throws ExternalReaderProcessException {
```

Comment:
These factories should be thread-safe.
```suggestion
    private synchronized ExternalModuleResolver getResolver() throws ExternalReaderProcessException {
```